### PR TITLE
fix(core): add `Send` + `Sync` impls for `RwLock`

### PR DESCRIPTION
**Pull request by zabackary**
## Describe the changes this PR makes. Why should it be merged?

`Send` and `Sync` implementations were not created for `RwLock`, making the lock essentially useless in Rust's type system.

This PR also updates the generic `T` in `RwLock` to be `?Sized` for more flexibility.

## Additional Context
- I haven't tested these changes
- semver: minor/patch (you pick)

### CODE REVIEW

**Review by zabackary — COMMENTED** [review]
I suppose this should technically be `feat(core`, but it probably doesn't matter too much.

File: packages/vexide-core/src/sync/rwlock.rs [hunk 1]
```diff
@@ -142,10 +142,13 @@ impl<'a, T> Future for RwLockWriteFuture<'a, T> {
 /// This type allows multiple readers or one writer at a time.
 ///
 /// This is different from a [`Mutex`](super::Mutex) because it allows for multiple readers at the same time.
-pub struct RwLock<T> {
+pub struct RwLock<T: ?Sized> {
     state: RwLockState,
     data: UnsafeCell<T>,
 }
+unsafe impl<T: ?Sized + Send> Send for RwLock<T> {}
+unsafe impl<T: ?Sized + Sync> Sync for RwLock<T> {}
```
Comment (Gavin-Niederman) [inline packages/vexide-core/src/sync/rwlock.rs:150]: This should be T: ?Sized + Send

**Review by Gavin-Niederman — CHANGES_REQUESTED** [review]

File: packages/vexide-core/src/sync/rwlock.rs (hunk 1 above)
Comment (doinkythederp) [reply packages/vexide-core/src/sync/rwlock.rs:150]: Can you make sure that’s consistent with the std RwLock? I’m pretty sure std::sync::RwLock needs T to be Sync to be Sync itself.

Comment (zabackary) [reply packages/vexide-core/src/sync/rwlock.rs:150]: https://doc.rust-lang.org/std/sync/struct.RwLock.html#impl-Send-for-RwLock%3CT%3E
https://doc.rust-lang.org/std/sync/struct.RwLock.html#impl-Sync-for-RwLock%3CT%3E

For the second one, it looks like `impl<T: ?Sized + Send + Sync> Sync for RwLock<T>` is the correct bound, maybe?

Comment (Gavin-Niederman) [reply packages/vexide-core/src/sync/rwlock.rs:150]: Yeah that's correct. I wasn't thinking about how RwLock works.

File: packages/vexide-core/src/sync/rwlock.rs [hunk 2]
```diff
@@ -142,10 +142,13 @@ impl<'a, T> Future for RwLockWriteFuture<'a, T> {
...
 /// This is different from a [`Mutex`](super::Mutex) because it allows for multiple readers at the same time.
-pub struct RwLock<T> {
+pub struct RwLock<T: ?Sized> {
     state: RwLockState,
     data: UnsafeCell<T>,
 }
+unsafe impl<T: ?Sized + Send> Send for RwLock<T> {}
+unsafe impl<T: ?Sized + Send + Sync> Sync for RwLock<T> {}
+
 impl<T> RwLock<T> {
```
Comment (Gavin-Niederman) [inline packages/vexide-core/src/sync/rwlock.rs:152]: This impl needs to also have a relaxed Sized bound.

**Review by Gavin-Niederman — CHANGES_REQUESTED** [review]

File: packages/vexide-core/src/sync/rwlock.rs (hunk 2 above)
Comment (Gavin-Niederman) [reply packages/vexide-core/src/sync/rwlock.rs:152]: You might need to put the new function in its own impl without a relaxed Sized bound

Comment (zabackary) [reply packages/vexide-core/src/sync/rwlock.rs:152]: Alternatively, the `new` function could be in this impl with a more restrictive `where T: Sized` clause

**Review by Tropix126 — APPROVED** [review]
LGTM

**Review by Gavin-Niederman — APPROVED** [review]
LGTM

**Pull request merged**